### PR TITLE
PARQUET-1810: [C++] Fix undefined behaviour on invalid enum values (OSS-Fuzz)

### PR DESCRIPTION
**Pull request by pitrou**
Should fix the following issues:
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20686
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20696
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20733
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20849
* https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=21010

### CODE REVIEW

File: cpp/src/parquet/thrift_internal.h [hunk 1]
```diff
@@ -63,29 +63,164 @@ using ::std::shared_ptr;
...
+};
+
+}  // namespace internal
+
+// Safe enum loader: will check for invalid enum value before converting
+
+template <typename ThriftType,
+          typename ParquetEnum =
+              typename internal::ThriftEnumTypeTraits<ThriftType>::ParquetEnum>
+inline typename ParquetEnum::type LoadEnumSafe(const ThriftType* in) {
```
Comment (pitrou) [inline cpp/src/parquet/thrift_internal.h]: Note: could perhaps rename this to `FromThrift(const ThriftType& in)` for convenience...

**Comment by pitrou** [comment]
cc @wesm @emkornfield

File: cpp/src/parquet/thrift_internal.h (hunk 1 above)
Comment (wesm) [reply cpp/src/parquet/thrift_internal.h]: We could. I don't have a strong opinion since it's an internal deserialization detail

**Review by wesm — APPROVED** [review]
+1. @emkornfield are you interested in taking a closer look at this? Can wait a day or two before merging

**Comment by pitrou** [comment]
Rebased.

**Comment by wesm** [comment]
Merging. We can always make follow up improvements

File: cpp/src/parquet/thrift_internal.h (hunk 1 above)
Comment (emkornfield) [reply cpp/src/parquet/thrift_internal.h]: I think the way the code reads now is fine as well.

**Comment by emkornfield** [comment]
Belated LGTM.

**Pull request closed without merging**